File: test_podcasts.js
const https = require('https');

// Helper to fetch text content from URL
function fetchUrl(url) {
    return new Promise((resolve, reject) => {
        const options = {
            headers: {
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
            }
        };
        https.get(url, options, (res) => {
            let data = '';
            res.on('data', (chunk) => { data += chunk; });
            res.on('end', () => { resolve(data); });
        }).on('error', (err) => { reject(err); });
    });
}

// Search Apple Podcast by Name
async function searchApplePodcast(name) {
    const url = `https://itunes.apple.com/search?term=${encodeURIComponent(name)}&country=tw&media=podcast&limit=1`;
    try {
        const resText = await fetchUrl(url);
        const data = JSON.parse(resText);
        if (data.resultCount > 0) {
            const result = data.results[0];
            return {
                podcastName: result.collectionName,
                appleId: result.collectionId,
                rssUrl: result.feedUrl,
                trackCount: result.trackCount
            };
        }
    } catch (e) {
        console.error(`搜尋 Apple Podcast 失敗 (${name}):`, e.message);
    }
    return null;
}

// Check eligibility from RSS XML
async function checkEligibility(rssUrl, monthsLimit = 6) {
    try {
        const xml = await fetchUrl(rssUrl);
        
        // Find Podcast Title
        const titleMatch = xml.match(/<title>(.*?)<\/title>/);
        const podcastTitle = titleMatch ? titleMatch[1].replace(/<!\[CDATA\[(.*?)\]\]>/g, '$1').trim() : 'Unknown';
        
        // Find all <item> blocks
        const items = [];
        const itemRegex = /<item>([\s\S]*?)<\/item>/g;
        let match;
        
        const current_time = new Date("2026-06-14T10:00:00Z"); // Set reference time
        const dateLimit = new Date(current_time.getTime() - (monthsLimit * 30 * 24 * 60 * 60 * 1000));
        
        while ((match = itemRegex.exec(xml)) !== null) {
            const itemContent = match[1];
            const titleM = itemContent.match(/<title>(.*?)<\/title>/);
            const pubDateM = itemContent.match(/<pubDate>(.*?)<\/pubDate>/);
            const enclosureM = itemContent.match(/<enclosure[^>]*url="([^"]+)"/);
            
            const title = titleM ? titleM[1].replace(/<!\[CDATA\[(.*?)\]\]>/g, '$1').trim() : '';
            const pubDateStr = pubDateM ? pubDateM[1] : '';
            const mp3Url = enclosureM ? enclosureM[1] : '';
            
            if (pubDateStr) {
                const pubDate = new Date(pubDateStr);
                if (!isNaN(pubDate.getTime()) && pubDate >= dateLimit) {
                    items.push({
                        title,
                        pubDate: pubDate.toISOString().split('T')[0],
                        mp3Url
                    });
                }
            }
        }
        
        return {
            title: podcastTitle,
            episodesCountPast6Months: items.length,
            eligible: items.length >= 12,
            sampleEpisodes: items.slice(0, 3)
        };
    } catch (e) {
        return { error: e.message };
    }
}

// Fetch Apple Podcast Reviews
async function getAppleReviews(appleId) {
    const url = `https://itunes.apple.com/tw/rss/customerreviews/id=${appleId}/json`;
    try {
        const resText = await fetchUrl(url);
        const data = JSON.parse(resText);
        const entries = data.feed?.entry || [];
        const reviews = [];
        
        // If it's a single entry, it's an object instead of array
        const entriesList = Array.isArray(entries) ? entries : [entries];
        
        for (const entry of entriesList) {
            if (entry.author && entry['im:rating']) {
                reviews.push({
                    author: entry.author.name.label,
                    rating: parseInt(entry['im:rating'].label),
                    title: entry.title.label,
                    content: entry.content.label
                });
            }
        }
        return {
            totalReviewsFound: reviews.length,
            sampleReviews: reviews.slice(0, 3)
        };
    } catch (e) {
        return { error: `取得 Apple 留言失敗: ${e.message}` };
    }
}

// Scrape Spotify rating count and value
async function getSpotifyRatings(spotifyUrl) {
    try {
        const html = await fetchUrl(spotifyUrl);
        
        // Search in JSON-LD scripts
        const ldRegex = /<script type="application\/ld\+json">([\s\S]*?)<\/script>/g;
        let match;
        while ((match = ldRegex.exec(html)) !== null) {
            try {
                const ld = JSON.parse(match[1].trim());
                if (ld.aggregateRating) {
                    return {
                        ratingValue: ld.aggregateRating.ratingValue,
                        ratingCount: ld.aggregateRating.ratingCount
                    };
                }
            } catch (e) {}
        }
        
        // Fallback: search for ratingValue / ratingCount keys in raw HTML strings
        const ratingM = html.match(/"ratingValue"\s*:\s*([0-9.]+)/);
        const countM = html.match(/"ratingCount"\s*:\s*([0-9,]+)/);
        if (ratingM && countM) {
            return {
                ratingValue: parseFloat(ratingM[1]),
                ratingCount: parseInt(countM[1].replace(/,/g, ''))
            };
        }
        
        // Search for Spotify rating display format: e.g. "4.9 (1.2k)" or "4.9 (1,234)"
        const textMatch = html.match(/([0-5]\.[0-9])\s+\(([0-9,kK\s]+)\)/);
        if (textMatch) {
            return {
                ratingValue: parseFloat(textMatch[1]),
                ratingCountRaw: textMatch[2].trim()
            };
        }
        
        return { status: "HTML中未找到 aggregateRating，可能因為未登入或客戶端渲染" };
    } catch (e) {
        return { error: `抓取 Spotify 失敗: ${e.message}` };
    }
}

async function main() {
    console.log("=================== 測試 1: 任性歐逆機智生活 ===================");
    const oniApple = await searchApplePodcast("任性歐逆機智生活");
    console.log("Apple 搜尋結果:\n", JSON.stringify(oniApple, null, 2));
    
    if (oniApple) {
        console.log("\n進行資格審查與集數抽樣...");
        const eligibility = await checkEligibility(oniApple.rssUrl);
        console.log("資格審查結果:\n", JSON.stringify(eligibility, null, 2));
        
        console.log("\n取得 Apple 聽眾留言...");
        const reviews = await getAppleReviews(oniApple.appleId);
        console.log("Apple 留言結果:\n", JSON.stringify(reviews, null, 2));
    }
    
    console.log("\n抓取 Spotify 評分人數...");
    const oniSpotify = await getSpotifyRatings("https://open.spotify.com/show/3RSITJFvOU7hy3VcDKYUBU");
    console.log("Spotify 評分結果:\n", JSON.stringify(oniSpotify, null, 2));
    
    console.log("\n=================== 測試 2: 科技領航家 朱楚文 ===================");
    const chuApple = await searchApplePodcast("科技領航家");
    console.log("Apple 搜尋結果:\n", JSON.stringify(chuApple, null, 2));
    
    if (chuApple) {
        console.log("\n進行資格審查與集數抽樣...");
        const eligibility = await checkEligibility(chuApple.rssUrl);
        console.log("資格審查結果:\n", JSON.stringify(eligibility, null, 2));
        
        console.log("\n取得 Apple 聽眾留言...");
        const reviews = await getAppleReviews("1485503209");
        console.log("Apple 留言結果:\n", JSON.stringify(reviews, null, 2));
    }
    
    console.log("\n抓取 Spotify 評分人數...");
    const chuSpotify = await getSpotifyRatings("https://open.spotify.com/show/7o50v1V5w4oNFRfH6Fnx4f");
    console.log("Spotify 評分結果:\n", JSON.stringify(chuSpotify, null, 2));
}

main();

File: test_podcasts.py
import os
import xml.etree.ElementTree as ET
import urllib.request
import json
import re
from datetime import datetime, timedelta
import email.utils

def search_apple_podcast(name):
    """
    使用 Apple Search API 搜尋 Podcast，取得 Apple ID 和 RSS URL
    """
    url = f"https://itunes.apple.com/search?term={urllib.parse.quote(name)}&country=tw&media=podcast&limit=1"
    try:
        req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
        with urllib.request.urlopen(req) as response:
            data = json.loads(response.read().decode('utf-8'))
        if data.get('resultCount', 0) > 0:
            result = data['results'][0]
            return {
                "podcast_name": result.get("collectionName"),
                "apple_id": result.get("collectionId"),
                "rss_url": result.get("feedUrl"),
                "track_count": result.get("trackCount")
            }
    except Exception as e:
        print(f"搜尋 Apple Podcast 失敗: {e}")
    return None

def check_eligibility_and_list_episodes(rss_url, months_limit=6):
    """
    讀取 RSS Feed，計算過去 6 個月內的總集數與清單
    """
    try:
        req = urllib.request.Request(rss_url, headers={'User-Agent': 'Mozilla/5.0'})
        with urllib.request.urlopen(req) as response:
            xml_data = response.read()
        
        root = ET.fromstring(xml_data)
        channel = root.find('channel')
        if channel is None:
            return None, "無法找到 RSS channel 標籤"
            
        podcast_title = channel.findtext('title', 'Unknown').strip()
        episodes_in_range = []
        
        # 2026-06-14 往回推 6 個月
        current_time = datetime(2026, 6, 14)
        date_limit = current_time - timedelta(days=months_limit * 30)
        
        for item in channel.findall('item'):
            title = item.findtext('title', '').strip()
            pub_date_str = item.findtext('pubDate', '')
            enclosure = item.find('enclosure')
            mp3_url = enclosure.get('url') if enclosure is not None else None
            
            pub_date = None
            if pub_date_str:
                try:
                    pub_tuple = email.utils.parsedate_tz(pub_date_str)
                    if pub_tuple:
                        pub_date = datetime.fromtimestamp(email.utils.mktime_tz(pub_tuple))
                except Exception:
                    pass
            
            if pub_date and pub_date >= date_limit:
                episodes_in_range.append({
                    "title": title,
                    "pub_date": pub_date.strftime("%Y-%m-%d"),
                    "mp3_url": mp3_url
                })
        
        eligible = len(episodes_in_range) >= 12
        return {
            "title": podcast_title,
            "episodes_count_past_6_months": len(episodes_in_range),
            "eligible": eligible,
            "episodes": episodes_in_range
        }, None
    except Exception as e:
        return None, str(e)

def get_apple_reviews(apple_id):
    """
    透過 Apple Customer Reviews RSS 取得留言數與評分
    """
    url = f"https://itunes.apple.com/tw/rss/customerreviews/id={apple_id}/json"
    try:
        req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
        with urllib.request.urlopen(req) as response:
            data = json.loads(response.read().decode('utf-8'))
        
        feed = data.get('feed', {})
        entries = feed.get('entry', [])
        
        # 如果只有一筆，iTunes 會直接回傳 object 而不是 list，需要包裝一下
        if isinstance(entries, dict):
            entries = [entries]
            
        reviews = []
        # entry 的第一筆通常是節目的基本資訊，而不是留言，所以要過濾
        for entry in entries:
            if 'author' in entry and 'im:rating' in entry:
                author = entry['author']['name']['label']
                rating = int(entry['im:rating']['label'])
                title = entry['title']['label']
                content = entry['content']['label']
                reviews.append({
                    "author": author,
                    "rating": rating,
                    "title": title,
                    "content": content
                })
        return {
            "total_reviews_found": len(reviews),
            "sample_reviews": reviews[:3] # 取前三筆展示
        }
    except Exception as e:
        return {"error": f"取得 Apple 留言失敗: {e}"}

def get_spotify_ratings(spotify_url):
    """
    嘗試爬取 Spotify 頁面中的評分和星等
    """
    try:
        req = urllib.request.Request(spotify_url, headers={'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'})
        with urllib.request.urlopen(req) as response:
            html = response.read().decode('utf-8')
        
        # 在 HTML 中搜尋類似 "4.8 (1,234)" 或 schema.org JSON-LD 中的 rating 資訊
        # Spotify 常在 script type="application/ld+json" 中放置 ratingValue 與 ratingCount
        json_ld_matches = re.findall(r'<script type="application/ld\+json">(.*?)</script>', html, re.DOTALL)
        for match in json_ld_matches:
            try:
                ld = json.loads(match.strip())
                # 有時候是 @type: Series 或 PodcastSeries
                if "aggregateRating" in ld:
                    rating = ld["aggregateRating"]
                    return {
                        "rating_value": rating.get("ratingValue"),
                        "rating_count": rating.get("ratingCount")
                    }
            except Exception:
                pass
        
        # 備用：正則表達式匹配
        # 例如："4.9 (1.2k)" 或 "4.9 (1,234)"
        # 尋找包含 rating 的 meta 或 text
        rating_match = re.search(r'ratingValue[":\s]+([0-9.]+)', html)
        count_match = re.search(r'ratingCount[":\s]+([0-9,]+)', html)
        if rating_match and count_match:
            return {
                "rating_value": float(rating_match.group(1)),
                "rating_count": int(count_match.group(1).replace(',', ''))
            }
            
        # 嘗試在普通文本中尋找
        # 例如 Spotify 上常見格式: 5.0 (13)
        span_match = re.search(r'([0-5]\.[0-9])\s+\(([0-9,kK\s]+)\)', html)
        if span_match:
            return {
                "rating_value": float(span_match.group(1)),
                "rating_count_raw": span_match.group(2).strip()
            }
            
    except Exception as e:
        return {"error": f"抓取 Spotify 失敗: {e}"}
    return {"status": "未在 HTML 中找到 rating 欄位，可能被 React 遮蔽"}

if __name__ == "__main__":
    print("=================== 測試 1: 任性歐逆機智生活 ===================")
    # 搜尋 Apple ID
    oni_apple = search_apple_podcast("任性歐逆機智生活")
    print("Apple 搜尋結果:", json.dumps(oni_apple, ensure_ascii=False, indent=2))
    
    if oni_apple:
        print("\n進行資格審查與集數抽樣...")
        info, err = check_eligibility_and_list_episodes(oni_apple['rss_url'])
        if info:
            print(f"節目名稱: {info['title']}")
            print(f"過去 6 個月集數: {info['episodes_count_past_6_months']}")
            print(f"符合評選資格 (>=12集): {'【是】' if info['eligible'] else '【否】'}")
            if info['episodes']:
                print("最新三集範例:")
                for ep in info['episodes'][:3]:
                    print(f" - [{ep['pub_date']}] {ep['title']}")
        else:
            print("審查錯誤:", err)
            
        print("\n取得 Apple 聽眾留言...")
        reviews = get_apple_reviews(oni_apple['apple_id'])
        print(json.dumps(reviews, ensure_ascii=False, indent=2))
        
    print("\n抓取 Spotify 評分人數...")
    oni_spotify = get_spotify_ratings("https://open.spotify.com/show/3RSITJFvOU7hy3VcDKYUBU")
    print("Spotify 評分結果:", json.dumps(oni_spotify, ensure_ascii=False, indent=2))
    
    print("\n=================== 測試 2: 科技領航家 朱楚文 ===================")
    chu_apple_id = "1485503209"
    # 也可以用搜尋
    chu_apple = search_apple_podcast("科技領航家")
    print("Apple 搜尋結果:", json.dumps(chu_apple, ensure_ascii=False, indent=2))
    
    if chu_apple:
        print("\n進行資格審查與集數抽樣...")
        info, err = check_eligibility_and_list_episodes(chu_apple['rss_url'])
        if info:
            print(f"節目名稱: {info['title']}")
            print(f"過去 6 個月集數: {info['episodes_count_past_6_months']}")
            print(f"符合評選資格 (>=12集): {'【是】' if info['eligible'] else '【否】'}")
            if info['episodes']:
                print("最新三集範例:")
                for ep in info['episodes'][:3]:
                    print(f" - [{ep['pub_date']}] {ep['title']}")
        else:
            print("審查錯誤:", err)
            
        print("\n取得 Apple 聽眾留言...")
        reviews = get_apple_reviews(chu_apple_id)
        print(json.dumps(reviews, ensure_ascii=False, indent=2))
        
    print("\n抓取 Spotify 評分人數...")
    chu_spotify = get_spotify_ratings("https://open.spotify.com/show/7o50v1V5w4oNFRfH6Fnx4f")
    print("Spotify 評分結果:", json.dumps(chu_spotify, ensure_ascii=False, indent=2))
